crypto/rsa/rsa_gen.c

```diff
@@ -85,62 +134,171 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
+             * prime should not be equal to p, q, r_3...
+             * (those primes prior to this one)
+             */
+            {
+                int j = 0;
+                BIGNUM *prev_prime = NULL;
+
+                for (j = 0; j < i; j++) {
+                    if (j == 0)
+                        prev_prime = rsa->p;
+                    else if (j == 1)
+                        prev_prime = rsa->q;
+                    else
+                        prev_prime = sk_RSA_PRIME_INFO_value(prime_infos,
+                                                             j - 2)->r;
+
+                    if (!BN_cmp(prime, prev_prime)) {
+                        goto redo;
+                    }
+                }
+            }
+            if (!BN_sub(r2, prime, BN_value_one()))
+                goto err;
+            if (!BN_gcd(r1, r2, rsa->e, ctx))
+                goto err;
+            if (BN_is_one(r1))
+                break;
+            if (!BN_GENCB_call(cb, 2, n++))
+                goto err;
+        }
+
+        bitse += bitsr[i];
+
+        /* calculate n immediately to see if it's sufficient */
+        if (i == 1) {
+            /* we get at least 2 primes */
+            if (!BN_mul(r1, rsa->p, rsa->q, ctx))
+                goto err;
+        } else if (i != 0) {
+            /* modulus n = p * q * r_3 * r_4 ... */
+            if (!BN_mul(r1, rsa->n, prime, ctx))
+                goto err;
+        } else {
+            /* i == 0, do nothing */
+            if (!BN_GENCB_call(cb, 3, i))
+                goto err;
+            continue;
+        }
+        nl = BN_num_bits(r1);
+        /*
+         * if |r1|, product of factors so far, is not long enough,
+         * or the first 4 bits are less than 0x9, re-generate the last prime.
+         *
+         * NOTE: This actually can't happen in two-prime case, because of
+         * the way factors are generated. But for multi-prime case, the
+         * modulus could start less than 0x9, which could be utilized to
+         * distinguish a multi-prime private key by using the modulus
+         * in a certificate.
+         */
+        if (!BN_rshift(r2, r1, nl - 4))
             goto err;
-    }
-    if (!BN_GENCB_call(cb, 3, 0))
-        goto err;
-    for (;;) {
-        do {
-            if (!BN_generate_prime_ex(rsa->q, bitsq, 0, NULL, NULL, cb))
+        bitst = BN_get_word(r2);
+
+        if (nl != bitse || bitst < 0x9) {
+            /*
+             * For keys with more than 4 primes, we attempt longer factor to
+             * meet length requirement.
+             *
+             * Otherwise, we just re-generate the prime with the same length.
+             *
+             * This strategy has the following goals:
+             *
+             * 1. 1024-bit factors are effcient when using 3072 and 4096-bit key
+             * 2. stay the same logic with normal 2-prime key
+             */
+            if (primes > 4)
+                adj += bitse - nl;
```

Comment:
If `nl == bitse` and `bitst == 0x8` this does not adjust `adj`
this may enter an endless loop.

Reply:
assuming you change above shift to `BN_rshift(r2, r1, bitse - 4)`,
I would suggest to increment adj if bitst<0x9 and to decrement if bitst>0xF
regardless of `primes > 4` test above,.

Reply:
.. or re-try a *limited* number of times with adj unchanged.

Reply:
Okay, I get your point. And `primes > 4` was discussed with @dot-asm in previous comments to make sure 'special' length of primes for 3072 and 4096 key (although we don't consider the key length yet in this PR), so it could be kept here...

Reply:
And also, I think there should be some comments in the code to explain why the check is `< 9 || > 0xf`, this piece of code is hard to read with out comments...

Reply:
yes, the key points from this discussion here should get into the comments.

Reply:
I agree that this test makes some sense, but it is too rigorous.

I see the following problem with prime#3, and prime#4:

It may happen that the p*q or p*q*r is very close to (9/16)*2^bitse,
so when you generate the next prime, the product will be with very
high likelihood less than (9/16)*2^bitse.  So I suggest you should
not re-generate more than 4 times with adj==0.

Furthermore I think that `adj should never be less than zero or greater than one`.
That should be possible to prove as a mathematical theorem.

I would like to have an assertion for that.

Reply:
Well, if you absolutely want to generate 4*1024 bit primes, you could try a few times
and start the whole search from scratch, because if it does not help to re-generate
the last prime then the product of all the previous primes was too low.